components/raftstore/src/store/peer.rs

```diff
@@ -4387,6 +4401,58 @@ where
             }
         }
     }
+
+    pub fn unsafe_recovery_wait_apply<T: Transport>(
+        &mut self,
+        ctx: &PollContext<EK, ER, T>,
+        counter: Arc<AtomicUsize>,
+    ) {
+        info!(
+            "unsafe recovery wait apply";
+            "target_index" => self.raft_group.store().commit_index(),
+            "task_counter" => counter.load(Ordering::Relaxed)
+        );
+
+        self.unsafe_recovery_state = Some(UnsafeRecoveryState::WaitApply {
+            target_index: self.raft_group.store().commit_index(),
+            task_counter: counter,
+        });
+        self.unsafe_recovery_maybe_finish_wait_apply(ctx, /*force=*/ false);
+    }
+
+    pub fn unsafe_recovery_maybe_finish_wait_apply<T: Transport>(
+        &mut self,
+        ctx: &PollContext<EK, ER, T>,
+        force: bool,
+    ) {
+        if let Some(unsafe_recovery_state) = self.unsafe_recovery_state.as_ref() {
+            debug!(
+                "unsafe recovery may be finish wait apply";
+                "applied" =>  self.raft_group.raft.raft_log.applied,
+                "force" => force
+            );
+            let UnsafeRecoveryState::WaitApply {
+                target_index,
+                task_counter,
+            } = unsafe_recovery_state;
+            if self.raft_group.raft.raft_log.applied >= *target_index || force {
+                info!(
+                    "unsafe recovery finish wait apply";
+                    "target_index" => target_index,
+                    "applied" =>  self.raft_group.raft.raft_log.applied,
+                    "force" => force,
+                    "counter" =>  task_counter.load(Ordering::Relaxed)
+                );
+                if task_counter.fetch_sub(1, Ordering::Relaxed) == 1 {
```

Comment:
Should be `SeqCst`.

Reply:
Memory order does not matter here right, since we only do one read-modify-write operation per thread?

Reply:
Sending control is also an atomic operation. Unless it doesn't show up in condition or there is further double check on the atomic, `Relaxed` should probably not be used.

Reply:
Control sending depends on the evaluation of the atomic operation. So the order is already assured?

Reply:
For example, what if the heartbeat is sent with stale region info? Besides, it's not a performance sensitive path, using `Relaxed` doesn't introduce any noticeable improvement but always bring safety concerns.

Reply:
Changed.

Reply:
I can see it's still Relaxed.

Reply:
Sorry, wrong place, changed now.